server/lib/s3backup/__init__.py

```diff
@@ -0,0 +1,130 @@
+    """
+    def __init__(self, name, md5):
+        self.name = name
+        self.md5 = md5
+
+    def __eq__(self, other):
+        return self.name == other.name and self.md5 == other.md5
+
+
+class S3Connector(object):
+    def __init__(self, access_key_id, secret_access_key, endpoint_url, bucket_name):
+        self.s3connector = boto3.client('s3',
+                                        aws_access_key_id='{}'.format(
+                                            access_key_id),
+                                        aws_secret_access_key='{}'.format(
+                                            secret_access_key),
+                                        endpoint_url='{}'.format(endpoint_url))
+
+    def list_objects(self, Bucket):
+        return self.s3connector.list_objects_v2(Bucket=Bucket)
+
+    def put_object(self, Bucket, Key, Body, ContentMD5):
+        return self.s3connector.put_object(Bucket=Bucket, Key=Key,
+                Body=Body, ContentMD5=ContentMD5)
+
+    def head_bucket(self, Bucket):
+        return self.s3connector.head_bucket(Bucket=Bucket)
+
+    def get_object(self, Bucket, Key):
+        return self.s3connector.get_object(Bucket=Bucket, Key=Key)
+
+
+class MockS3Connector(object):
+    """
+    The mock object is used for unit testing. It provides a "connector" to the backend service that is implemented using the local filesystem, rather than dispatching to the real S3 backend service.
+    """
+
+    def __init__(self, access_key_id, secret_access_key, endpoint_url, bucket_name):
+        self.path = endpoint_url
+        self.bucket_name = bucket_name
+
+    def list_objects(self, Bucket):
+        ob_dict = {}
+        bucketpath = os.path.join(self.path, Bucket)
+        result_list = glob.glob(os.path.join(bucketpath, "*/*.tar.xz"))
+        result_name_list = []
+        for i in result_list:
+            with open(i, 'rb') as f:
+                data = f.read()
+                md5 = hashlib.md5(data).hexdigest()
+            result_name_list.append({'ETag': '"{}"'.format(md5),
+                                     'Key': os.path.relpath(i, start=bucketpath)})
+        ob_dict['Contents'] = result_name_list
+        ob_dict['ResponseMetadata'] = {'HTTPStatusCode': 400}
+        return ob_dict
+
+    def put_object(self, Bucket=None, Key=None, Body=None, ContentMD5=None):
+        md5_hex_value = hashlib.md5(Body.read()).hexdigest()
+        md5_base64_value = (base64.b64encode(
+                        bytes.fromhex(md5_hex_value))).decode()
+        if md5_base64_value == ContentMD5:
+            test_controller = Key.split("/")[0]
+            try:
+                os.mkdir("{}/{}/{}".format(self.path,
+                                           self.bucket_name, test_controller))
+            except FileExistsError:
+                # directory already exists, ignore
+                pass
+            with open('{}/{}/{}.tar.xz'.format(self.path, self.bucket_name, Key), 'wb') as f:
```

Comment:
I thought I already flagged this as wrong: the key already contains '.tar.xz' so you are duplicating it. Fixing that will have repercussions on the unit tests too.